python/GafferImageUI/OpenColorIOContextUI.py

```diff
@@ -0,0 +1,208 @@
+#        copyright notice, this list of conditions and the following
+#        disclaimer in the documentation and/or other materials provided with
+#        the distribution.
+#
+#      * Neither the name of John Haddon nor the names of
+#        any other contributors to this software may be used to endorse or
+#        promote products derived from this software without specific prior
+#        written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+import imath
+
+import Gaffer
+import GafferUI
+import GafferImage
+
+##########################################################################
+# Metadata
+##########################################################################
+
+Gaffer.Metadata.registerNode(
+
+	GafferImage.OpenColorIOContext,
+
+	"description",
+	"""
+	Creates Gaffer context variables which define the OpenColorIO config
+	to be used by upstream nodes. This allows different configs to be used
+	in different contexts.
+	""",
+
+	"layout:section:Settings.Variables:collapsed", False,
+
+	plugs = {
+
+		"in" : [
+
+			"plugValueWidget:type", "",
+
+		],
+
+		"out" : [
+
+			"plugValueWidget:type", "",
+
+		],
+
+		"config" : [
+
+			"description",
+			"""
+			The OpenColorIO config to use.
+			""",
+
+			"nodule:type", "",
+			"plugValueWidget:type", "GafferUI.LayoutPlugValueWidget",
+			"layoutPlugValueWidget:orientation", "horizontal",
+
+		],
+
+		"config.enabled" : [
+
+			"description",
+			"""
+			Enables the `config.value` plug, allowing the OpenColorIO config
+			to be specified.
+			""",
+
+			"nodule:type", "",
+			"boolPlugValueWidget:displayMode", "switch",
+			"label", "",
+
+		],
+
+		"config.value" : [
+
+			"description",
+			"""
+			Specifies the OpenColorIO config to be used.
+			""",
+
+			"nodule:type", "",
+			"label", "",
+			"layout:activator", lambda plug : plug.parent()["enabled"].getValue(),
+			"plugValueWidget:type", "GafferUI.PresetsPlugValueWidget",
+			"presetsPlugValueWidget:allowCustom", True,
+
+			"preset:$OCIO", "",
+			"preset:ACES CG", "ocio://cg-config-v1.0.0_aces-v1.3_ocio-v2.1",
+			"preset:ACES Studio", "ocio://studio-config-v1.0.0_aces-v1.3_ocio-v2.1",
```

Comment:
In the spirit of "preserve OCIO's names", should we consider using the more verbose names for the Studio and CG config presets? I'm not convinced `cg-config-v1.0.0_aces-v1.3_ocio-v2.1` is a particularly great user-facing name, but `ACES CG` is pretty close to `ACEScg`. I'll see if anything strikes me in the morning...

Reply:
> I'm not convinced cg-config-v1.0.0_aces-v1.3_ocio-v2.1 is a particularly great user-facing name

It's horrendous! Took me a while to dig up the [semantics](https://github.com/AcademySoftwareFoundation/OpenColorIO-Config-ACES/releases/tag/v1.0.0) again :

>studio-config-v1.0.0_aces-v1.3_ocio-v2.1.ocio
|-----|-------------|---------|---------|---|
|Type | Colorspaces | ACES    | Profile |   | 
>Where:
> Type : The type of the config, e.g. Reference, CG or Studio.
    Colorspaces : The colorspace definitions versions, semantically versioned.
    ACES : Supported [aces-dev](https://github.com/ampas/aces-dev) version.
    Profile : Supported OCIO profile, greater versions will enable more features, e.g. ACES 1.3 Reference Gamut Compression for 2.1 profile.

I suppose someone could care about any one of those version numbers at some point in time, but right now they're the only sensible ones available I think.

> ACES CG is pretty close to ACEScg

I'm not sure if you're proposing that `ACEScg` would be a better and more official name, or that `ACES CG` is a bad name because it is close to `ACEScg` and `ACEScg` is a colorspace name rather than a config name. I don't think `ACEScg` is a good name for the preset, because it's a colorspace, and I suppose there is a small potential for confusion with a colorspace name.

I sought guidance in the [ACES UX guidelines](https://www.dropbox.com/s/v5tghz6wl2629nf/TB-2014-002.pdf?dl=0) but they don't seem to specify anything useful here - I think the CG vs Studio distinction is something that OCIO introduced rather than being in the ACES spec. They _do_ say the version number should be displayed though.

The [OpenColorIO documentation](https://opencolorio.readthedocs.io/en/latest/configurations/aces_studio.html) refers to the config in headings as `ACES Studio Config` and `ACES CG Config`, and then give the long strings in the body text.

I propose the following :

- We don't use the three-versioned full string in the preset name - it's too complex for most folks (and me unless I wander round the internet for a bit to find the documentation of the semantics).
- We do include the ACES version number, per the UX guidelines, because it's the most useful user-facing version, and because it helps disambiguates "ACES CG" from the similar colorspace name. So that would be "ACES CG 1.3" and "ACES Studio 1.3".
- The `Custom` option provides easy access to the full string for folks who care about the gory details and might want to modify it.

Reply:
I did the "1.3" thing in https://github.com/GafferHQ/gaffer/pull/5324/commits/c5a29646f4c23bc09b7621ac15587768be08d595, but can revisit if you'd prefer something else.

Reply:
How about the following? We move the "1.3" next to ACES as it's specifying the ACES version rather than the config version, and then use the OCIO documentation's header terms for the configs with a "-" separating the two.

`ACES 1.3 - CG Config`
`ACES 1.3 - Studio Config`

I think this is pretty close to what we were converging on while discussing the potential options offline. It is a mishmash of the pertinent information, but a more friendly preset name than  `cg-config-v1.0.0_aces-v1.3_ocio-v2.1`.

Reply:
Sounds good to me. I did that in https://github.com/GafferHQ/gaffer/pull/5324/commits/a034065dce658eb72e7b2aa4653656e39fe77889, rebasing to get the fix from https://github.com/GafferHQ/gaffer/pull/5326 and squashing in all the fixups.